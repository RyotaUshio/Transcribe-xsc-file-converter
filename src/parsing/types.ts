export type MarkerType =
  | { name: "section"; relativeScope: 3 }
  | { name: "measure"; relativeScope: 2 }
  | { name: "beat"; relativeScope: 1 }
export type NumSubdividionsData = {
  type: "same_as_previous_marker" | "constant"
  numSubdivisionsBetweenThisMarkerAndNext?: number
}
export type LabelInfo = null | { label: string; labelIsAutoNamed: boolean }
export type TimestampAndParts = { string: string; hoursPart: number; minutesPart: number; secondsPart: number }

export interface ParseData {
  soundFile: {
    filename?: string
    filePath?: string
    audioDurationInSeconds?: number
  }
  markers: {
    list: Array<{
      timestamp: TimestampAndParts
      markerType: MarkerType
      label: LabelInfo
      numSubdivisionsBetweenThisAndNextMarker: NumSubdividionsData
    }>
    autonumbering: {
      byMarkerType: { section?: boolean; measure?: boolean; beat?: boolean }
      numberContinuously?: boolean
    }
  }
}

export interface ParseState {
  data: ParseData
  currentSection: null | string
}

export type LineParser = (line: string, currentState: ParseState) => ParseState
