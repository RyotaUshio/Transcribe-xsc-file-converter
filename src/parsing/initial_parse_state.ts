import { ParseState } from "./types"

export const initialParseState: ParseState = {
  data: {
    soundFile: {},
    markers: {
      list: [],
      autonumbering: { byMarkerType: {} },
    },
  },
  currentSection: null,
}
