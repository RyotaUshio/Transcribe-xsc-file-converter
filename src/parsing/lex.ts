export function splitIntoLines(string: string): Array<string> {
  return string.split(/\r\n|\n|\r/)
}

export function splitLineIntoParts(line: string): Array<string> {
  const split = line.split(",")
  return split.map(unescapeLinePart)
  // Splitting on commas before unescaping is okay because the escaped line parts never contain commas.
  // This is because commas are escaped as `\C`, not as `\,`.
}

/**
 * Returns a string with only the following differences:
 * - `\\` is interpreted as `\`
 * - `\C` (with a non-escaped backslash) is interpreted as `,`
 */
export function unescapeLinePart(linePart: string): string {
  // Minor performance note: this code uses repeated .push on Array followed by .join because I assume it is faster than repeated += on string. I did not benchmark to confirm that assumption.

  const escapedChars: Array<string> = []
  let nextCharIsEscape = false

  for (const char of linePart) {
    if (nextCharIsEscape) {
      if (char === "\\") {
        escapedChars.push("\\")
      } else if (char === "C") {
        escapedChars.push(",")
      } else {
        // For example, "\n" or "\r" or "\t" or "\\".
        escapedChars.push("\\", char)
      }
      nextCharIsEscape = false
    } else if (char === "\\") {
      nextCharIsEscape = true
    } else {
      escapedChars.push(char)
    }
  }

  if (nextCharIsEscape) {
    throw new Error(
      "couldn’t interpret line part ending in an escaping ‘\\’ with no following character: '" + linePart + "'"
    )
  }

  return escapedChars.join("")
}

export function getKeyOfLine(line: string): string {
  // Reusing `splitLineIntoParts` is simpler than trying to match only until the first comma, and it ensures we won’t have to change this function if we later realize some escape character can affect splitting.
  const parts = splitLineIntoParts(line)
  return parts[0]
}

export function lineIsBlank(line: string): boolean {
  return /^\s*$/.test(line)
}
