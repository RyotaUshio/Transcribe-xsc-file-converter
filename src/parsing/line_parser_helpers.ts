import { ParseState } from "./types"

export function assertCurrentSection(expectedSection: string, currentState: ParseState): void {
  if (currentState.currentSection !== expectedSection) {
    throw new Error(
      `found a field expected to be in the ${expectedSection} section in the ${currentState.currentSection} section instead; change code to support that`
    )
  }
}
