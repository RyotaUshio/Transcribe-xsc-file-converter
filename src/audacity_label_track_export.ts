import { ParseData, TimestampWithPreparsedParts } from "./parsing/types"

// reference links:
// Label Track file format details: https://manual.audacityteam.org/man/importing_and_exporting_labels.html
// about Label Tracks in Audacity: https://manual.audacityteam.org/man/label_tracks.html
// about Audacity: https://www.audacityteam.org/

function timestampToSeconds(timestampData: TimestampWithPreparsedParts): number {
  const { numericParts } = timestampData
  return numericParts.hoursPart * (60 * 60) + numericParts.minutesPart * 60 + numericParts.secondsPart
}

export function toAudacityLabelTrackFormat(parseData: ParseData): string {
  const outputLines: Array<string> = []

  const markerList = parseData.markers.list
  markerList.forEach((item) => {
    const lineParts = []

    const timestampInSeconds = timestampToSeconds(item.timestamp)
    // make a point label by giving it the same start time and end time
    lineParts.push(timestampInSeconds) // start time
    lineParts.push(timestampInSeconds) // end time

    // When importing markers without labels, Audacity supports either line format: "1\t1" (missing tab-separated label field) or "1\t1\t" (label field present but empty).
    // I chose to export the second format, a present but empty label field, because it’s more likely to avoid bugs in hypothetical other programs that understand Audacity’s label track format.
    let labelName = item.label ? item.label.name : ""
    lineParts.push(labelName)

    outputLines.push(lineParts.join("\t"))
  })

  if (outputLines.length === 0) {
    return ""
    // returning "\n" would be invalid; that causes Audacity to show an error message about a missing label
  } else {
    return outputLines.join("\n") + "\n"
  }
}
