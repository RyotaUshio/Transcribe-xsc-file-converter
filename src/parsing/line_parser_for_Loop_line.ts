import { produce } from "immer"
import { splitLineIntoParts } from "./lex"
import { assertCurrentSection } from "./line_parser_helpers"
import { ParseState } from "./types"
import { timestampToObject } from "./timestamp"

export function parseLoop(line: string, currentState: ParseState): ParseState {
    assertCurrentSection("Loops", currentState)

    // L,1,1,3415162,35466,,White,,0:01:11.149200,0:00:00.738881
    const [
        loopType, // always "L"
        loopIndexStr, 
        _unknown1, // "1" or "0"
        _unknown2,
        _unknown3,
        _unknown4,
        color,
        _unknown5,
        loopStartStr,
        loopLengthStr,
    ] = splitLineIntoParts(line)

    return produce(currentState, (draftState) => {
        draftState.data.loops.list.push({
            index: parseInt(loopIndexStr),
            color,
            startTimestamp: timestampToObject(loopStartStr),
            lengthTimestamp: timestampToObject(loopLengthStr),
        })
    })
}