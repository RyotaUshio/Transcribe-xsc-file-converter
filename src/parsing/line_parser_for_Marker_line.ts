import produce from "immer"
import { splitLineIntoParts } from "./lex"
import { NumSubdividionsData, LabelInfo, ParseState, TimestampAndParts } from "./types"
import { assertCurrentSection } from "./line_parser_helpers"

export function parseMarker(line: string, currentState: ParseState): ParseState {
  const [
    markerTypeChar,
    _frameOrByteOrSampleOrSomething, // This is (seconds_elapsed * 44.1) for one test file. 44.1 is the “sampling rate” of that audio file, 44,100 Hz, divided by 1000. Therefore, this value means ???
    labelIsAutoNamedStr,
    labelName,
    numSubdivisionsStr,
    timestamp,
  ] = splitLineIntoParts(line)

  assertCurrentSection("Markers", currentState)

  return produce(currentState, (draftState) => {
    draftState.data.markers.list.push({
      timestamp: timestampToObject(timestamp),
      markerType: markerTypeCharToObject(markerTypeChar),
      label: labelInfoToObjectOrNull(labelName, labelIsAutoNamedStr),
      numSubdivisionsBetweenThisAndNextMarker: numSubdivisionsStringToObject(numSubdivisionsStr),
    })
  })
}

function numSubdivisionsStringToObject(numSubdivisionsStr: string): NumSubdividionsData {
  const numSubdivisionsValue = parseInt(numSubdivisionsStr, 10)

  if (numSubdivisionsValue === 0) {
    return { type: "same_as_previous_marker" }
  } else if (numSubdivisionsValue === 1) {
    return {
      type: "constant",
      numSubdivisionsBetweenThisMarkerAndNext: 0,
    }
  } else {
    return {
      type: "constant",
      numSubdivisionsBetweenThisMarkerAndNext: numSubdivisionsValue,
    }
  }
}

function timestampToObject(timestampStr: string): TimestampAndParts {
  const [hours, minutes, seconds] = timestampStr.split(":")
  return {
    string: timestampStr,
    hoursPart: parseInt(hours, 10),
    minutesPart: parseInt(minutes, 10),
    secondsPart: parseFloat(seconds),
  }
}

type MarkerTypeDetailsFromChar = {
  S: { name: "section"; relativeScope: 3 }
  M: { name: "measure"; relativeScope: 2 }
  B: { name: "beat"; relativeScope: 1 }
}
const markerTypeDetailsFromChar: MarkerTypeDetailsFromChar = {
  S: { name: "section", relativeScope: 3 },
  M: { name: "measure", relativeScope: 2 },
  B: { name: "beat", relativeScope: 1 },
}
// If you want to simplify the above duplication, read https://www.typescriptlang.org/docs/handbook/advanced-types.html and see if anything applies. Searches didn’t bring up anything more specific than that page.
function markerTypeCharToObject(markerTypeChar: string) {
  if (markerTypeChar === "S" || markerTypeChar === "M" || markerTypeChar === "B") {
    return markerTypeDetailsFromChar[markerTypeChar]
  } else {
    throw new Error(`invalid marker type "#{markerTypeChar}" passed to markerTypeCharToObject`)
  }
}

function labelInfoToObjectOrNull(name: string, isAutoNamedStr: string): LabelInfo {
  const isAutoNamed = isAutoNamedStr === "1" ? true : false
  if (isAutoNamed && name === "") {
    return null
  } else {
    return { name, isAutoNamed }
  }
}
