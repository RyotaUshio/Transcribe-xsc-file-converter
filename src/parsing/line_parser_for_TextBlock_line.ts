import { produce } from "immer";
import { ParseState } from "./types";
import { assertCurrentSection } from "./line_parser_helpers";
import { splitLineIntoParts } from "./lex";
import { timestampToObject } from "./timestamp";

export function parseTextBlock(line: string, currentState: ParseState): ParseState {
    assertCurrentSection("TextBlocks", currentState);

    const [
        textBlockType, // always "T" 
        _unknown, // always "-1" for test files 
        height, 
        width, 
        color, 
        timestamp, 
        text, 
        fitWidthStr,
    ] = splitLineIntoParts(line);

    return produce(currentState, (draftState) => {
        draftState.data.textBlocks.list.push({
            timestamp: timestampToObject(timestamp),
            color,
            text,
            height: parseInt(height),
            width: parseInt(width),
            fitWidth: fitWidthStr === "1",
        });
    });
}

export function parseTextBlockFont(line: string, currentState: ParseState): ParseState {
    assertCurrentSection("TextBlocks", currentState);

    const [_key, _unknown, fontSize, fixedWidthFontName, proportionalWidthFontName] = splitLineIntoParts(line);

    return produce(currentState, (draftState) => {
        draftState.data.textBlocks.font = {
            fixedWidthFontName: fixedWidthFontName,
            proportionalWidthFontName: proportionalWidthFontName,
            size: parseInt(fontSize),
        };
    });
}
