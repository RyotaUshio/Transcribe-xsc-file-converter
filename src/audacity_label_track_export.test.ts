import { toAudacityLabelTrackFormat } from "./audacity_label_track_export"
import { parse } from "./parsing/parser"
import { readAllOfAFile } from "./io_utils"
import { ParseData } from "./parsing/types"

test("parsed .xsc data with no timestamps can be converted to Audacity Label Track format", async () => {
  const parseDataWithOneMarker: ParseData = {
    soundFile: {},
    markers: {
      list: [],
      autonumbering: { byMarkerType: {} },
    },
  }
  const labelTrackFormat = toAudacityLabelTrackFormat(parseDataWithOneMarker)
  expect(labelTrackFormat).toEqual("")
})

test("parsed .xsc data with a single late, unlabeled timestamp can be converted to Audacity Label Track format", async () => {
  const parseDataWithOneMarker: ParseData = {
    soundFile: {},
    markers: {
      list: [
        {
          label: null,
          markerType: { name: "beat", relativeScope: 1 },
          numSubdivisionsBetweenThisAndNextMarker: { type: "same_as_previous_marker" },
          timestamp: {
            hoursPart: 1,
            minutesPart: 59,
            secondsPart: 4.157,
            string: "1:59:04.157",
          },
        },
      ],
      autonumbering: { byMarkerType: {} },
    },
  }
  const labelTrackFormat = toAudacityLabelTrackFormat(parseDataWithOneMarker)
  expect(labelTrackFormat).toEqual("7144.157\t7144.157\t\n")
})

test("parsed .xsc data with a single labeled timestamp can be converted to Audacity Label Track format", async () => {
  const parseDataWithOneMarker: ParseData = {
    soundFile: {},
    markers: {
      list: [
        {
          label: {
            name: "start of the good part",
            isAutoNamed: false,
          },
          markerType: { name: "beat", relativeScope: 1 },
          numSubdivisionsBetweenThisAndNextMarker: { type: "same_as_previous_marker" },
          timestamp: {
            hoursPart: 0,
            minutesPart: 0,
            secondsPart: 3,
            string: "0:00:03.000",
          },
        },
      ],
      autonumbering: { byMarkerType: {} },
    },
  }
  const labelTrackFormat = toAudacityLabelTrackFormat(parseDataWithOneMarker)
  expect(labelTrackFormat).toEqual("3\t3\tstart of the good part\n")
})

test("the file ‘Windows ME Startup.xsc’ can be converted to Audacity Label Track format", async () => {
  const fileContents = await readAllOfAFile("tests/relatively few markers/Windows ME Startup.xsc")
  const parsed = parse(fileContents)

  const labelTrackFormat = toAudacityLabelTrackFormat(parsed)

  // use snapshot assertion to avoid managing trailing tabs in code files
  expect(labelTrackFormat).toMatchSnapshot()
})
