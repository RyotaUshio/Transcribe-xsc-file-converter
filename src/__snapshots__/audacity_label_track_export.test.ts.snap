// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the file ‘Windows ME Startup.xsc’ can be converted to Audacity Label Track format 1`] = `
"0.167	0.167	A1
1.039	1.039	
1.961	1.961	A2
2.678	2.678	
3.343	3.343	
4	4	
4.637	4.637	
"
`;
